File: Blind 75 LeetCode/Chahat Rajput/CourseSchedule/course_schedule.cpp
class Solution {
public:
    bool topologicalSortCheck(unordered_map<int,vector<int>>&adj,int n,vector<int>&indegree){
        queue<int>q;
        int count=0;//kitne node visit kr liye
        for(int i=0;i<n;i++){
            if(indegree[i] == 0 ){
                count++;//kitne nodes mene visit kr liye use add kr diya queue me
                q.push(i);
            }
        }

        while(!q.empty()){
            int u=q.front();
            q.pop();

            for(int &v:adj[u]){
                indegree[v]--;

                if(indegree[v] == 0){
                    count++;
                    q.push(v);
                }
            }
        }
        if(count ==n){//able to visit all nodes
            return true;
        }
        return false;//means there was a cycle and I couldn't complete allcourses
    }
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        //Just to find that is there any cycle in graph then no topological sort will be created so in thta wayy we tackel this problem
        //Just to check Cycle detection in graph
         
         unordered_map<int,vector<int>>adj;
         vector<int>indegree(numCourses,0);//kahn's algorithm

         for(auto &vec:prerequisites){
             int a=vec[0];
             int b=vec[1];

             //pehle b fir  a b-->a
             adj[b].push_back(a);

             //arrow b se a me ja rha he
             indegree[a]++;
         }
         //if cycle is present,not possible
        return topologicalSortCheck(adj,numCourses,indegree);
    }
};

File: Blind 75 LeetCode/Subhangi dhasmana/Course Schedule/Course Schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indegree = [0] * numCourses
        adj = [[] for x in range(numCourses)]
        
        for prereq in prerequisites:
            adj[prereq[1]].append(prereq[0])
            indegree[prereq[0]] += 1

        queue = []
        for i in range(numCourses):
            if indegree[i] == 0:
                queue.append(i)
        
        visited = 0
        while queue:
            node = queue.pop(0)
            visited += 1
            for neighbor in adj[node]:
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    queue.append(neighbor)
        
        return numCourses == visited